### ISSUE TITLE
Adding Unit Summary Entity - Discussion

### ISSUE THREAD

**Issue by seanmalbert** [bloom-housing/bloom#1588]
# Overview
Detroit has a need to enter and store unit summaries directly. On a longer time scale, Bloom does as well, but there is a more immediate desire to move the processing of transformUnits to upserting a listing and storing the data in a table, so we can read from that directly instead of processing it on request.

Ideally, we can align on the same structures, so we'll need to see what Detroit's requirements are.

# Columns
This is a little difficult to pinpoint what would make sense for Bloom, since transform units returns six different structures:
- byUnitTypeAndRent
- byUnitType
- byReservedType
- byAMI
- hmi

I have more investigation to do to figure out a structure that would best work for all of these and I'll add that to a comment soon, but don't want to block Detroit.

# Implementation
@willrlin mentioned the idea of using a derived table for unit summary. That could possibly work for Bloom's case of using the units table to create the derivatives, but how would that work for Detroit directly writing to a table?

We can create a feature flag feature to determine whether the unit summary table should be directly writable or synced from units. In both cases though, we want to read from the the unit summary table(s).

# Current Considerations
Our team is currently creating a way to override Ami Chart values, by annual income percentage. These values will be taken into account when processing the unit summary data.

**Comment by willrlin** [comment]
@abbiefarr @anders-schneider please use this issue to discuss our needs and align on this work.

**Comment by anders-schneider** [comment]
Hey @seanmalbert, thanks for starting a thread about this. This all sounds good to @abbiefarr and me: always read from the new unit summary table, and have a feature flag to determine whether that table is writable directly or is derived from the units table.

If we can agree on the structure of the new unit summary table, I think that would unblock us so we can create the new table and start on a couple different workstreams. Abbie and I poked around [`UnitsSummarized`](https://github.com/bloom-housing/bloom/blob/3fd11f849cc84f133d544b52037661560f57c2df/backend/core/src/units/types/units-summarized.ts#L11) and the [`UnitSummary`](https://github.com/bloom-housing/bloom/blob/3fd11f849cc84f133d544b52037661560f57c2df/backend/core/src/units/types/unit-summary.ts#L9)’s that are included within it, and have ideas about a possible structure for the new unit summary table that would offer the same granularity as the existing unitsSummarized field. Proposed columns:

- Unit type (string)
- Rent (string)
- Rent as percentage of income (number)
- Reserved type (string)
- AMI percentage (string)
- Minimum income minimum (string)
- Minimum income maximum (string)
- Occupancy minimum (number)
- Occupancy maximum (number)
- Area minimum (number)
- Area maximum (number)
- Floor minimum (number)
- Floor maximum (number)
- Total count (number)
- Total available count (number)

For the Detroit use case, many of these fields will be null. I’m not sure if your codebase uses all the different fields within UnitSummary (I assume so?), but if not we might be able to trim them from `UnitSummary` before starting any of this work (and reduce the number of proposed columns for the new unit summary table).

We have more thoughts about how to approach this work, but as a first step, we’d love to hear your thoughts on that proposed table structure!

**Comment by seanmalbert** [comment]
Hey @anders-schneider ,

Thanks for putting that together. I think this works for our needs too, though I'm wondering if we want to reference ids where applicable instead of saving the string. For example, take Unit Type, which is a related entity to Units. If we save the string value in UnitSummary, which is the label of the unit type, and then edit the unit type label, UnitSummary would not reflect that, unless we updated all rows where the old value exists with the new one.

I understand that this isn't a concern really right now, because there isn't a UI to manage these "dropdown" options, but eventually there will be. But I also like that it's a string, as it reduces the amount of joins or lookups we have to do. So for `Unit Type` and `Reserved Type` it'd be nice to have an id (string) column. They can be nullable for Detroit's case, and we can use them to update already computed unit summaries.

With jurisdictions like Alameda, we do need all of those values, but we can create different "views" as @avaleske suggests [here](https://github.com/bloom-housing/bloom/discussions/1563#discussioncomment-1080536), so Detroit could get the "slim" set of unit data.

**Comment by anders-schneider** [comment]
Thanks @seanmalbert and agreed that it makes sense to consider referencing ids instead of storing string labels.

> But I also like that it's a string, as it reduces the amount of joins or lookups we have to do. So for Unit Type and Reserved Type it'd be nice to have an id (string) column. They can be nullable for Detroit's case, and we can use them to update already computed unit summaries.

Do I understand correctly that you're suggesting we have two columns for Unit Type: one with the string label and the other with an id pointing to a row in the unit_types table? And in the Detroit case, we might store `null`s in the unit_type_id column and avoid joining with the unit_type table? (And the same pattern for Reserved Type.)

**Comment by seanmalbert** [comment]
@anders-schneider ,
I was suggesting that, but thinking about it again, that's probably not a good idea for the long term and I'd prefer to just store the reference ids.

**Comment by anders-schneider** [comment]
Yeah agreed - let's just store reference ids! So I believe the columns would be:

- Unit type id (foreign key, referencing `UnitType` entity)
- Unit accessibility priority type (foreign key, referencing `UnitAccessibilityPriorityType` entity)
- Rent (string)
- Rent as percentage of income (number)
- AMI percentage (string)
- Minimum income minimum (string)
- Minimum income maximum (string)
- Occupancy minimum (number)
- Occupancy maximum (number)
- Area minimum (number)
- Area maximum (number)
- Floor minimum (number)
- Floor maximum (number)
- Total count (number)
- Total available count (number)

I removed `Reserved Type` because it looks like you all are moving away from setting that at the unit level (https://github.com/bloom-housing/bloom/pull/1540), and we also don't expect to need anything like that at the unit level.

And I added a foreign key referencing `UnitAccessibilityPriorityType`, since `UnitsSummarized` includes an array of them (missed that in our first pass of proposed columns).

@abbiefarr is working on a PR to create this new table - that may be a good place to figure out further details?

**Comment by seanmalbert** [comment]
This looks great, @anders-schneider and thanks for catching the reserved and priority type updates. Yes, I think the PR will work fine for figuring out the rest.

Thanks @anders-schneider and @abbiefarr!
